rinad/src/ipcm/ipcm.cc

```diff
@@ -1769,4 +1786,266 @@ void IPCManager_::io_loop(){
+			continue;
+		}
+
+		cpsets[ps->name] = CatalogPsInfo(*ps, plit);
+	}
+}
+
+// Helper function used by load_by_template()
+void Catalog::psinfo_from_psconfig(list< rina::PsInfo >& psinfo_list,
+				   const string& component,
+				   const rina::PolicyConfig& pconfig)
+{
+	if (pconfig.name_ != string()) {
+		psinfo_list.push_back(rina::PsInfo(pconfig.name_,
+						   component,
+						   pconfig.version_));
+	}
+}
+
+int Catalog::load_by_template(Addon *addon, unsigned int ipcp_id,
+			      const rinad::DIFTemplate *t)
+{
+	list< rina::PsInfo > required_policy_sets;
+
+	// Collect into a list all the policy sets specified
+	// by the template
+	psinfo_from_psconfig(required_policy_sets, "rmt",
+			     t->rmtConfiguration.policy_set_);
+
+	psinfo_from_psconfig(required_policy_sets, "pff",
+			     t->rmtConfiguration.pft_conf_.policy_set_);
+
+	psinfo_from_psconfig(required_policy_sets, "enrollment-task",
+			     t->etConfiguration.policy_set_);
+
+	psinfo_from_psconfig(required_policy_sets, "security-manager",
+			     t->secManConfiguration.policy_set_);
+
+	psinfo_from_psconfig(required_policy_sets, "flow-allocator",
+			     t->faConfiguration.policy_set_);
+
+	psinfo_from_psconfig(required_policy_sets, "namespace-manager",
+			     t->nsmConfiguration.policy_set_);
+
+	psinfo_from_psconfig(required_policy_sets, "resource-allocator",
+			     t->raConfiguration.pduftg_conf_.policy_set_);
+
+	psinfo_from_psconfig(required_policy_sets, "routing",
+			     t->routingConfiguration.policy_set_);
+
+	// Load all the policy sets in the list
+	for (list<rina::PsInfo>::iterator i=required_policy_sets.begin();
+			i != required_policy_sets.end(); i++) {
+		int ret = load_policy_set(addon, ipcp_id, *i);
+
+		if (ret) {
+			LOG_WARN("Failed to load policy-set %s/%s",
+				  i->app_entity.c_str(), i->name.c_str());
+		}
+	}
+
+	return 0;
+}
+
+int Catalog::load_policy_set(Addon *addon, unsigned int ipcp_id,
+			     const rina::PsInfo& psinfo)
+{
+	Promise promise;
+	string plugin_name;
+
+	{
+		// Perform a first lookup (under read lock) to see if we
+		// can find a suitable plugin to load
+		rina::ReadScopedLock wlock(rwlock);
+
+		if (policy_sets.count(psinfo.app_entity) == 0) {
+			LOG_WARN("Catalog does not contain any policy-set "
+				 "for component %s",
+				 psinfo.app_entity.c_str());
+			return -1;
+		}
+
+		map<string, CatalogPsInfo>& cmap = policy_sets[psinfo.app_entity];
+
+		if (cmap.count(psinfo.name) == 0) {
+			LOG_WARN("Catalog does not contain a policy-set "
+				 "called %s for component %s",
+				 psinfo.name.c_str(),
+				 psinfo.app_entity.c_str());
+			return -1;
+		}
+
+		if (cmap[psinfo.name].plugin->second.loaded) {
+			// Nothing to do, we have already loaded this.
+			return 0;
+		}
+
+		plugin_name = cmap[psinfo.name].plugin->second.name;
+	}
+
+	// Perform the plugin loading - a blocking and possibly
+	// asynchronous operation - out of the catalog lock
+	if (IPCManager->plugin_load(addon, &promise, ipcp_id,
```

Comment:
I am very confused; it seems that the responsabilities of the IPCM and the Catalogue are not properly split. Calling the IPCM to load the plugin seems not the responsability of the Catalog.

My thoughts are:

* Catalogue: should solely have the "list" of available policy sets and policies (and the relation to policy set(s)), and its details (path, characteristics..).
* IPCM: in charge of consulting the catalogue, get the info, and load or unload the plugins.
* Who keeps track of the loaded or unloaded plugins? To me this is an IPCM class responsability.

My suggestion is to convert this method into a method that returns an exception or an `std::list<Plugin>`, so that the IPCM can call its internal methods to load.